src/languageSetup.ts

```diff
@@ -81,6 +83,10 @@ export async function activateLanguageServer(context: vscode.ExtensionContext, s
     }));
 
     await languageClient.onReady();
+
+    languageClient.onNotification("kotlin/buildOutputLocationSet", (buildOutputLocation: string) => kotlinApi.setBuildOutputPath(buildOutputLocation));
```

Comment:
Could we move the notification definition to `lspExtensions`?

Reply:
Also, I would like to make sure that this doesn't break any LSP clients that don't support our extension. With non-standard requests it's usually not a problem since the client simply doesn't send them, but notifications could be an issue. Perhaps the client should advertise a custom capability/something else in the initialize request so the server can safely send the notification? Or perhaps we could use a pull model where the client extension requests the `buildOutputLocation` directly from the language server though a request like `kotlin/buildOutputLocation` (this might require exposing the `LanguageClient` in some form of wrapper instead of the current `ExtensionAPI`?).

Reply:
I think this prevents the scenario you're talking about: https://github.com/fwcd/kotlin-language-server/pull/334/files#diff-90db0a634202af6b5c845208ba6e9bf93cbfc975e8dc8af32d7e81ddb640c0daR10-R14

With that said, I should probably catch the exception here: https://github.com/fwcd/kotlin-language-server/pull/334/files#diff-a47c02e7183139961f7073dc40f6e30139192d3429adc87d076155776c90ba27R136

I think this should completely prevent the notification from ever breaking the client

Reply:
That's good, but I still think that relying on exception handling shouldn't be our only measure here. Clients react to unsupported requests in different/unspecified ways and not every client might do so cleanly. For example, some clients might crash instead of sending out a JSON-RPC error, potentially making the language server unusable.

Reply:
Perhaps we're talking about different things? What I linked was all from the server. As far as I can understand, if the client does not support the notification, LSP4J will use the default I specified (i.e., it will throw an exception on the server), which means the notification would never even reach the client, right?

Reply:
I don't think the server knows about the methods a client supports and it will try to send the notification in any case (the default implementation shouldn't be invoked at all, since the interface is automatically implemented by the client proxy, it might even throw a different exception than `UnsupportedOperationException` and instead something JSON-RPC-related), I might be wrong though.

Reply:
Still, I would ideally like to have a solution that either uses a request or negotiates with the server before sending out non-standard notifications. If we were to use a request for this, it could even be asynchronous (possibly first resolving once the class path is ready or similar).

Reply:
I see your point. I'll make it a request then, since it's quite simple to implement.